pkg/name/name.go

```diff
@@ -0,0 +1,251 @@
+		return featureNode.Value
+	}
+	if componentNode == nil {
+		return featureNode.Value
+	}
+	return componentNode.Value
+}
+
+// Namespace returns the namespace for the component. It follows these rules:
+// 1. If CustomPackagePath is unset, log and error and return the empty string.
+// 2. If the feature and component namespaces are unset, return CustomPackagePath.
+// 3. If the feature namespace is set but component name is unset, return the feature namespace.
+// 4. Otherwise return the component namespace.
+func Namespace(featureName string, componentName ComponentName, installSpec *v1alpha2.IstioControlPlaneSpec) string {
+	defaultNamespaceI, found, err := GetFromStructPath(installSpec, "CustomPackagePath")
+	if !found {
+		log.Error("can't find any default for CustomPackagePath")
+		return ""
+	}
+	if err != nil {
+		log.Error(err.Error())
+		return ""
+
+	}
+	defaultNamespace, ok := defaultNamespaceI.(string)
+	if !ok {
+		log.Errorf("CustomPackagePath has bad type %T, expect string", defaultNamespaceI)
+		return ""
+	}
+
+	featureNamespace := defaultNamespace
+	featureNodeI, found, err := GetFromStructPath(installSpec, featureName+"Components.Namespace")
+	if err != nil {
+		log.Error(err.Error())
+		return featureNamespace
+	}
+	if found && featureNodeI != nil {
+		featureNamespace, ok = featureNodeI.(string)
+		if !ok {
+			log.Errorf("feature %s namespace has bad type %T, expect string", featureName, featureNodeI)
+			return defaultNamespace
+		}
+		if featureNamespace == "" {
+			featureNamespace = defaultNamespace
+		}
+	}
+
+	componentNamespace := featureNamespace
+	componentNodeI, found, err := GetFromStructPath(installSpec, featureName+".Components."+string(componentName)+".Common.Namespace")
+	if err != nil {
+		log.Error(err.Error())
+		return featureNamespace
+	}
+	if !found {
+		return featureNamespace
+	}
+	if componentNodeI == nil {
+		return featureNamespace
+	}
+	componentNamespace, ok = componentNodeI.(string)
+	if !ok {
+		log.Errorf("component %s enabled has bad type %T, expect string", componentName, componentNodeI)
+		return featureNamespace
+	}
+	if componentNamespace == "" {
+		return featureNamespace
+	}
+	return componentNamespace
+}
+
+// GetFromStructPath returns the value at path from the given node, or false if the path does not exist.
+// Node and all intermediate along path must be type struct ptr.
+func GetFromStructPath(node interface{}, path string) (interface{}, bool, error) {
+	return getFromStructPath(node, util.PathFromString(path))
+}
+
+// getFromStructPath is the internal implementation of GetFromStructPath which recurses through a tree of Go structs
+// given a path. It terminates when the end of the path is reached or a path element does not exist.
+func getFromStructPath(node interface{}, path util.Path) (interface{}, bool, error) {
+	kind := reflect.TypeOf(node).Kind()
+	var structElems reflect.Value
+	switch kind {
+	case reflect.Map, reflect.Slice:
+		if len(path) != 0 {
```

Comment:
I may understand it wrong, but why len(path)!=0 indicates unsupported leaf type for Map or Slice